contents/elm/Elm.md

```diff
@@ -0,0 +1,256 @@
+
+## What is Elm?
+
+[Elm](https://elm-lang.org/) is a statically typed, functional programming language that compiles to JavaScript and promises no runtime exceptions. You can use it to build reliable and maintainable applications for the web.
+
+
+## Why Learn Elm?
+
+Languages like Haskell and OCaml from the [ML family](https://en.wikipedia.org/wiki/ML_(programming_language)) have been well known to be maintainable and mostly error-free in practice. Their compilers catch bugs well before they can make it into production. However, these functional programming languages are also well known to have a steep learning curve.
+
+JavaScript, the standard language for web development, has a rather low learning curve but is considerably more difficult to maintain. To solve this, people write tests, add type annotations to JavaScript, and even try to adopt functional programming patterns like immutability to increase maintanability.
+
+Elm tries to be the best of both worlds. Its functional core provides maintainability, and it tries to greatly reduce the barriers to learning functional programming by simplifying concepts and having a really pleasant compiler.
+
+### Benefit: Handling All Cases
+Like many functional programming languages, you get the "if it compiles it works" feeling in Elm.
+
+If your code can produce more than one result (e.g. converting a string to an integer can either fail or succeed), Elm's compiler forces you to account for all cases in order to compile.
+
+---
+
+Let's say you have an array, and you try to access its 10th element. What happens if there is no such element?
+
+Some languages like JavaScript will give you `undefined`, and your code might very well continue to execute and only error out in a completely separate function, making it a pain to debug.
+
+Other languages, like Java and Haskell, will throw a form of `IndexOutOfBoundsException`. That's ideal for you as a developer since you know exactly at which line this error occurs.
+
+However, this doesn't make for a good user experience. What if you forget to test an edge case?
+
+---
+
+In Elm, you will **never** run into these errors.
+
+Elm forces you to handle both cases of the array access in your code:
+- the element exists
+- the element doesn't exist.
+
+If you don't handle them, your code won't compile!
+
+
+### Benefit: Friendly Compiler
+
+Elm's compiler strives to be as friendly as possible, and offers as many hints as it can to help you fix errors. If nothing else, you should try Elm just to see how great error messages can be.
+
+Here's a quick look at some of them:
+
+- Cyclic dependencies:
+```
+-- IMPORT CYCLE ----------------------------------------------------------------
+
+Your module imports form a cycle:
+
+    ┌─────┐
+    │    Bar
+    │     ↓
+    │    Foo
+    └─────┘
+
+Learn more about why this is disallowed and how to break cycles
+here:<https://elm-lang.org/0.19.0/import-cycles>
+```
```

Comment:
Should these output examples have line numbers?
On a related note, should we make line numbering default only if the code black has a language specifier? Is that even possible?

Reply:
> Should these output examples have line numbers?
> On a related note, should we make line numbering default only if the code black has a language specifier? Is that even possible?

I'll hide them, don't think they're useful here.

I thought about that, but it would be rather confusing. Sometimes code blocks have line numbers and sometimes they do not depending on whether there is a langauge.

Reply:
> I thought about that, but it would be rather confusing. Sometimes code blocks have line numbers and sometimes they do not depending on whether there is a langauge.

Well, it's not entirely random and there is a good rationale. Might was users some work. It's nice when things 'just work' without the user having to do some additional configurations.

Reply:
Of course, if we do that, we also have to add a way to 'enable' line numbers too. So, that's another thing to consider on the cost side.